File: Euclid/Messages/Incoming/MOVESTUFF.cs
using Euclid.Game;
using Euclid.Messages.Outgoing;
using Euclid.Network.Streams;

namespace Euclid.Messages.Incoming
{
    class MOVESTUFF : IMessageEvent
    {
        public void Handle(Player player, Request request)
        {
            int itemId = int.Parse(request.GetArgument(0));

            if (player.RoomEntity.Room == null)
                return;

            Room room = player.RoomEntity.Room;

            if (room == null)
                return;

            Item item = room.ItemManager.GetItem(itemId);

            if (item == null || item.Data.OwnerId != player.Details.Id) // TODO: Staff check
                return;


            int x = int.Parse(request.GetArgument(1));
            int y = int.Parse(request.GetArgument(2));
            int rotation = int.Parse(request.GetArgument(3));

            var oldPosition = item.Position.Copy();

            if ((oldPosition.X == x &&
                oldPosition.Y == y &&
                oldPosition.Rotation == rotation) || !item.IsValidMove(item, room, x, y, rotation))
            {
                if (new Position(x, y) != item.Position)
                {
                    player.Send(new ACTIVEOBJECT_UPDATE(item));
                    return;
                }
            }

            if (rotation % 2 != 0 || rotation < 0 || rotation > 6)
                rotation = 0;

            room.FurnitureManager.MoveItem(item, new Position
            {
                X = x,
                Y = y,
                Rotation = rotation
            });
        }
    }
}

File: euclid/src/main/java/euclid/messages/incoming/MOVESTUFF.java
package euclid.messages.incoming;

import euclid.game.Player;
import euclid.game.item.Item;
import euclid.game.pathfinder.Position;
import euclid.game.room.Room;
import euclid.messages.IMessageEvent;
import euclid.messages.outgoing.ACTIVEOBJECT_UPDATE;
import euclid.network.streams.Request;

public class MOVESTUFF implements IMessageEvent {

    @Override
    public void handle(Player player, Request request) {
        int itemId = Integer.parseInt(request.getArgument(0));

        if (player.getRoomEntity().getRoom() == null) {
            return;
        }

        Room room = player.getRoomEntity().getRoom();

        if (room == null) {
            return;
        }

        Item item = room.getItemManager().getItem(itemId);

        if (item == null || item.getData().getOwnerId() != player.getDetails().getId()) {
            return;
        }

        int x = Integer.parseInt(request.getArgument(1));
        int y = Integer.parseInt(request.getArgument(2));
        int rotation = Integer.parseInt(request.getArgument(3));

        Position oldPosition = item.getPosition().copy();

        if ((oldPosition.getX() == x &&
                oldPosition.getY() == y &&
                oldPosition.getRotation() == rotation) || !item.isValidMove(item, room, x, y, rotation)) {
            if (!new Position(x, y).equals(item.getPosition())) {
                player.send(new ACTIVEOBJECT_UPDATE(item));
                return;
            }
        }

        if (rotation % 2 != 0 || rotation < 0 || rotation > 6) {
            rotation = 0;
        }

        room.getFurnitureManager().moveItem(item, new Position(x, y, rotation));
    }
}
